cmd/vocab-service/neo4j.go

```diff
@@ -0,0 +1,331 @@
+			OR any(syn in c.synonyms where lower(syn) =~ {pattern})
+		RETURN c.id, v.id, c.label, c.code, c.synonyms
+	`)
+	if err != nil {
+		return err
+	}
+
+	rows, err := stmt.QueryNeo(map[string]interface{}{
+		"pattern": pattern,
+		"vocab":   vocab,
+	})
+	if err != nil {
+		return err
+	}
+	defer rows.Close()
+
+	for {
+		vals, _, err := rows.NextNeo()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return err
+		}
+
+		c := &Class{
+			ID:    vals[0].(string),
+			Vocab: vals[1].(string),
+			Label: vals[2].(string),
+			Code:  vals[3].(string),
+		}
+
+		if vals[4] != nil {
+			for _, v := range vals[4].([]interface{}) {
+				c.Synonyms = append(c.Synonyms, v.(string))
+			}
+		}
+
+		select {
+		case <-cxt.Done():
+			return nil
+
+		case res <- c:
+		}
+	}
+
+	return nil
+}
+
+// Get a class by code.
+func (s *service) Get(vocab, code string) (*Class, error) {
+	if vocab == "" {
+		return nil, errors.New("vocab cannot be empty")
+	}
+
+	if code == "" {
+		return nil, errors.New("code cannot be empty")
+	}
+
+	conn, err := s.pool.OpenPool()
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	stmt, err := conn.PrepareNeo(`
+		MATCH (v:Vocabulary {id: {vocab}})<-[:classOf]-(c:Class {code: {code}})
+		RETURN c.id, v.id, c.label, c.code, c.synonyms
+		LIMIT 1
+	`)
+	if err != nil {
+		return nil, err
+	}
+
+	rows, err := stmt.QueryNeo(map[string]interface{}{
+		"vocab": vocab,
+		"code":  code,
+	})
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	vals, _, err := rows.NextNeo()
+	if err == io.EOF {
+		return nil, nil
+	} else if err != nil {
+		return nil, err
+	}
+
+	c := &Class{
+		ID:    vals[0].(string),
+		Vocab: vals[1].(string),
+		Label: vals[2].(string),
+		Code:  vals[3].(string),
+	}
+
+	if vals[4] != nil {
+		for _, v := range vals[4].([]interface{}) {
+			c.Synonyms = append(c.Synonyms, v.(string))
+		}
+	}
+
+	return c, nil
+}
+
+// Checks whether a set of codes exist.
+func (s *service) Validate(vocab string, codes []string) ([]bool, error) {
+	if vocab == "" {
+		return nil, errors.New("vocab cannot be empty")
+	}
+
+	if len(codes) == 0 {
+		return nil, nil
+	}
+
+	conn, err := s.pool.OpenPool()
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	stmt, err := conn.PrepareNeo(`
+		MATCH (:Vocabulary {id: {vocab}})<-[:classOf]-(c:Class {code: {code}})
+		RETURN 1
+		LIMIT 1
+	`)
+	if err != nil {
+		return nil, err
+	}
+
+	params := map[string]interface{}{
+		"code":  "",
+		"vocab": vocab,
+	}
+
+	bools := make([]bool, len(codes))
+
+	for i, code := range codes {
+		params["code"] = code
+
+		rows, err := stmt.QueryNeo(params)
+		if err != nil {
+			return nil, err
+		}
+
+		_, _, err = rows.NextNeo()
+		rows.Close()
+
+		if err == io.EOF {
+			continue
+		} else if err != nil {
+			return nil, err
+		}
+
+		bools[i] = true
+	}
+
+	return bools, nil
+}
+
+func (s *service) traversal(cxt context.Context, query, vocab, code string, res chan<- *Class) error {
```

Comment:
Document with a comment please. Something like "Execute the given query using the passed vocab and code as parameters and return all resulting classes on the res channel", maybe?